crates/ty_python_semantic/src/types.rs

```diff
@@ -1166,11 +1165,76 @@ impl<'db> Type<'db> {
         }
     }
 
+    /// Return `true` if subtyping is reflexive for this type; `T <: T` is always true for any `T`
+    /// of this type.
+    ///
+    /// This is true for fully static types, but also for some types that may not be fully static.
+    /// For example, a `ClassLiteral` may inherit `Any`, but its subtyping is still reflexive.
+    ///
+    /// This method may have false negatives, but it should not have false positives.
+    fn subtyping_is_reflexive(self) -> bool {
+        match self {
+            Type::Never
+            | Type::FunctionLiteral(..)
+            | Type::BoundMethod(_)
+            | Type::WrapperDescriptor(_)
+            | Type::MethodWrapper(_)
+            | Type::DataclassDecorator(_)
+            | Type::DataclassTransformer(_)
+            | Type::ModuleLiteral(..)
+            | Type::IntLiteral(_)
+            | Type::BooleanLiteral(_)
+            | Type::StringLiteral(_)
+            | Type::LiteralString
+            | Type::BytesLiteral(_)
+            | Type::SpecialForm(_)
+            | Type::KnownInstance(_)
+            | Type::AlwaysFalsy
+            | Type::AlwaysTruthy
+            | Type::PropertyInstance(_)
+            // might inherit `Any`, but subtyping is still reflexive
+            | Type::ClassLiteral(_) => true,
+            _ => false,
```

Comment:
I'd prefer to have an exhaustive match here, so that we don't forget to update this method if we add other variants in the future where subtyping is reflexive